darts/timeseries.py

```diff
@@ -1006,98 +1378,73 @@ def update(self,
 
         Returns
         -------
         TimeSeries
             A new multivariate TimeSeries instance.
         """
-        raise_if_not((self.time_index() == other.time_index()).all(), 'The indices of the two TimeSeries instances '
+        raise_if_not(self.has_same_time_as(other), 'The indices of the two TimeSeries instances '
                      'must be equal', logger)
+        raise_if_not(self.n_samples == other.n_samples, 'Two series can be stacked only if they '
+                                                        'have the same number of samples.', logger)
+
+        other_xa = other.data_array(copy=False)
+        if other_xa.dims[0] != self._time_dim:
+            new_other_xa = xr.DataArray(other_xa.values,
+                                        dims=self._xa.dims,
+                                        coords={self._time_dim: self._time_index, DIMS[1]: other.components})
+        else:
+            new_other_xa = other_xa
 
-        new_dataframe = pd.concat([self.pd_dataframe(), other.pd_dataframe()], axis=1)
-        return TimeSeries(new_dataframe, self.freq_str())
+        new_xa = xr.concat((self._xa, new_other_xa), dim=DIMS[1])
 
-    @property
-    def width(self) -> int:
-        """
-        Returns
-        -------
-        int
-            The number of components (univariate time series) of the current TimeSeries instance.
-        """
-        return self._df.shape[1]
+        # we call the factory method here to disambiguate column names if needed.
+        return TimeSeries.from_xarray(new_xa, fill_missing_dates=False)
 
-    def univariate_component(self, index: int) -> 'TimeSeries':
+    def univariate_component(self, index: Union[str, int]) -> 'TimeSeries':
         """
         Retrieves one of the components of the current TimeSeries instance
         and returns it as new univariate TimeSeries instance.
 
         Parameters
         ----------
         index
-            An zero-indexed integer indicating which component to retrieve.
+            An zero-indexed integer indicating which component to retrieve. If components have names,
+            this can be a string with the component's name.
 
         Returns
         -------
         TimeSeries
             A new univariate TimeSeries instance.
         """
-
-        raise_if_not(index >= 0 and index < self.width, 'The index must be between 0 and the number of components '
-                     'of the current TimeSeries instance - 1, {}'.format(self.width - 1), logger)
-
-        return TimeSeries.from_series(self.pd_dataframe().iloc[:, index], freq=self.freq_str())
+        if isinstance(index, int):
+            new_xa = self._xa.isel(component=index).expand_dims(DIMS[1], axis=1)
+        else:
+            new_xa = self._xa.sel(component=index).expand_dims(DIMS[1], axis=1)
+        return TimeSeries(new_xa)
 
     def add_datetime_attribute(self, attribute: str, one_hot: bool = False) -> 'TimeSeries':
```

Comment:
Here it says attribute must be string (str) but in the doc string under Parameters it says pd.DatetimeIndex.

Reply:
You're right, I removed the type annotation here because I'm not 100% sure.